### PR TITLE
Add various performance tests.

### PR DESCRIPTION
**Pull request by pranavbhandari24**
Add the following performance tests:
 * Spanner to Text
 * Streaming Data generator to Spanner 
 * Spanner to Avro
 * Avro to Bigtable

### CODE REVIEW

**Review by Polber — DISMISSED** [review]

File: it/src/main/java/com/google/cloud/teleport/it/DataGenerator.java
Comment (pranavbhandari24) [reply it/src/main/java/com/google/cloud/teleport/it/DataGenerator.java]: Done

File: it/src/main/java/com/google/cloud/teleport/it/PerformanceBenchmarkingBase.java
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/PerformanceBenchmarkingBase.java:205]: awaitility.org could be an alternative so we keep polling for metrics instead of hardcoding X minutes. but it's ok to ignore for now (or maybe add a TODO with intentions)

File: v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtablePerformanceIT.java
Comment (bvolpato) [inline v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtablePerformanceIT.java:58]: something to think / longer term - do you think we can try to mimic TemplateTestBase to stage those templates on demand? Or make that an option/property? It would be good to validate benchmark across changes of a template, instead of using a fixed baseline (both are good, but we can optionally set `-DspecPath=gs://dataflow-templates/latest/GCS_Avro_to_Cloud_Bigtable` like handled here)

File: v2/streaming-data-generator/src/test/java/com/google/cloud/teleport/v2/templates/StreamingDataGeneratorPerformanceIT.java
Comment (bvolpato) [inline v2/streaming-data-generator/src/test/java/com/google/cloud/teleport/v2/templates/StreamingDataGeneratorPerformanceIT.java:267]: just curious - would somehow externalize QPS/numWorkers/maxNumWorkers help? or use parameterized tests to have different sets of tests?

File: v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtablePerformanceIT.java
Comment (bvolpato) [inline v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtablePerformanceIT.java:111]: nit: how realistic is to have 1 million/s? Is it really the intention to cause backpressure on the DataGenerator? maybe having some constants (e.g. BACKPRESSURE_QPS, HEAVY_QPS) would help understanding

**Review by bvolpato — DISMISSED** [review]
Thanks Pranav.

I added a few nits/suggestions but no concerns. Feel free to skip or address separately. 

Note: Feel free to `Google LGTM` too, I'm just leaving as is so @Polber can take a second look.

**Review by Polber — APPROVED** [review]
I agree with all of @bvolpato comments, but they can definitely be TODO's for another PR. Great work!

**Comment by Polber** [comment]
The issue that is preventing the unit tests from passing should be resolved in #561 

Once that PR is merged, the checks can be re-run.

File: it/src/main/java/com/google/cloud/teleport/it/PerformanceBenchmarkingBase.java
Comment (pranavbhandari24) [reply it/src/main/java/com/google/cloud/teleport/it/PerformanceBenchmarkingBase.java:205]: This is a bit complex since the metrics are available but they are late. Sleeping for 4 minutes guarantees that we get the most accurate metrics. However, I agree that we shouldn't hardcode X minutes.

I will look into this and fix this later. Added a todo to do this.

File: v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtablePerformanceIT.java
Comment (pranavbhandari24) [reply v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtablePerformanceIT.java:58]: I'm not sure we should do that since we are planning on moving the IT module to the beam repo soon? PerformanceBenchmarkingBase will be used for non-templates tests too

File: v2/streaming-data-generator/src/test/java/com/google/cloud/teleport/v2/templates/StreamingDataGeneratorPerformanceIT.java
Comment (pranavbhandari24) [reply v2/streaming-data-generator/src/test/java/com/google/cloud/teleport/v2/templates/StreamingDataGeneratorPerformanceIT.java:267]: Having parameterized tests for QPS/num workers/max num workers won't give us much information, but will increase our tests by a lot. Also, most of the time the pipeline autoscales down/up in a matter of a few minutes.

File: v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtablePerformanceIT.java
Comment (pranavbhandari24) [reply v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtablePerformanceIT.java:111]: 1 million QPS is very unrealistic since we have a message limit of only 35 million. The intention is to cause backpressure on the data generator. The current implementation of Streaming data generator template creates stragglers since the fanout of `GenerateSequence.from(...).to(...)` is very high. This leads to costly and slow processing in the pipeline. I plan on fixing this by adding a `Reshuffle` before `Generate Fake Messages` step in the pipeline, but we are currently blocked by https://github.com/apache/beam/issues/25140.

Having a high QPS and a high number of starting workers reduces the fanout thereby improving processing time and cost (cost reduces by more than 100% for generating 10gb of data) 

Note: This issue doesn't occur for streaming pipelines from Streaming Data generator template, only for batch pipelines.

Comment (pranavbhandari24) [reply v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtablePerformanceIT.java:111]: Adding constants for clarity is a good idea. I'll do this in a separate PR since I will need to make changes to other performance tests too.

**Review by bvolpato — APPROVED** [review]

File: it/src/main/java/com/google/cloud/teleport/it/launcher/PipelineLauncher.java
Comment (bvolpato) [inline it/src/main/java/com/google/cloud/teleport/it/launcher/PipelineLauncher.java:179]: nit: This feels more like a `Map<String, Object>` (to allow any value type) rather than a specific `T`. I feel <T> without specifying a type will reach the same outcome, but it may raise some raw type warnings.

File: v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtableLT.java
Comment (bvolpato) [inline v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtableLT.java:96]: nit:
I found that using `if (bigtableResourceManager != null)` around cleanups are a good practice for those tests.

If the setup/ResourceManager fails for any reason, this will be invoked and can throw a NullPointerException, so you end up getting some extra errors/red herrings on the output.

File: v1/src/test/java/com/google/cloud/teleport/templates/PubsubToBigQueryLT.java
Comment (bvolpato) [inline v1/src/test/java/com/google/cloud/teleport/templates/PubsubToBigQueryLT.java:153]: `>= NUM_MESSAGES` is safer here due to Beam/Dataflow not offering exactly-once guarantees in some cases.

File: v2/googlecloud-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/templates/pubsubtotext/PubsubToTextLT.java
Comment (bvolpato) [inline v2/googlecloud-to-googlecloud/src/test/java/com/google/cloud/teleport/v2/templates/pubsubtotext/PubsubToTextLT.java:73]: Command explaining magic number = ~10gb

File: v1/src/test/java/com/google/cloud/teleport/templates/PubsubToBigQueryLT.java
Comment (bvolpato) [inline v1/src/test/java/com/google/cloud/teleport/templates/PubsubToBigQueryLT.java:105]: +1, much better with params adder.

**Review by bvolpato — APPROVED** [review]
LGTM. Added a few nit comments, feel free to ignore.

File: it/src/main/java/com/google/cloud/teleport/it/launcher/PipelineLauncher.java
Comment (pranavbhandari24) [reply it/src/main/java/com/google/cloud/teleport/it/launcher/PipelineLauncher.java:179]: I agree, changed it to `Map<String, Object>`

File: v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtableLT.java
Comment (pranavbhandari24) [reply v1/src/test/java/com/google/cloud/teleport/bigtable/AvroToBigtableLT.java:96]: Added checks in all performance tests

File: v1/src/test/java/com/google/cloud/teleport/templates/PubsubToBigQueryLT.java
Comment (pranavbhandari24) [reply v1/src/test/java/com/google/cloud/teleport/templates/PubsubToBigQueryLT.java:153]: Done

**Pull request merged**
